pyro/distributions/spanning_tree.py

```diff
@@ -112,20 +112,21 @@ def log_partition_function(self):
         # use a Cholesky decomposition to compute the log determinant.
         # See https://en.wikipedia.org/wiki/Kirchhoff%27s_theorem
         V = self.num_vertices
-        grid = make_complete_graph(V)
-        shift = self.edge_logits.detach().max()
-        edge_probs = (self.edge_logits - shift).exp()
-        adjacency = torch.zeros(V, V, dtype=edge_probs.dtype)
-        adjacency[grid[0], grid[1]] = edge_probs
-        adjacency[grid[1], grid[0]] = edge_probs
-        laplacian = adjacency.sum(-1).diag() - adjacency
+        v1, v2 = make_complete_graph(V).unbind(0)
+        logits = self.edge_logits.new_full((V, V), -math.inf)
+        logits[v1, v2] = self.edge_logits
+        logits[v2, v1] = self.edge_logits
+        log_diag = logits.logsumexp(-1)
+        # Numerically stabilize so that laplacian has 1's on the diagonal.
+        shift = 0.5 * log_diag
+        laplacian = torch.eye(V) - (logits - shift - shift[:, None]).exp()
```

Comment:
looks right to me. what's the deal with the truncation?

Reply:
That's just how the theorem goes: you delete any single column and row, then compute the determinant.
See https://en.wikipedia.org/wiki/Kirchhoff%27s_theorem#An_example_using_the_matrix-tree_theorem